File: 648-replace-words/replace-words.py
class Solution:
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        # Build Trie from dictionary
        root = TrieNode()
        for word in dictionary:
            node = root
            for ch in word:
                if ch not in node.children:
                    node.children[ch] = TrieNode()
                node = node.children[ch]
            node.endOfWord = True

        # Replace words in sentence
        words = sentence.split()
        for i in range(len(words)):
            node = root
            for j in range(len(words[i])):
                if words[i][j] in node.children:
                    node = node.children[words[i][j]]
                    if node.endOfWord:
                        words[i] = words[i][:j+1]
                        break
                else:
                    break

        return ' '.join(words)

class TrieNode:
    def __init__(self):
        self.children = {}
        self.endOfWord = False

File: 648-replace-words/replace-words.js
class TrieNode {
    constructor() {
        this.children = {};
        this.isEnd = false;
    }
}

class Trie {
    constructor() {
        this.root = new TrieNode();
    }

    insert(word) {
        let node = this.root;
        for (let ch of word) {
            if (!node.children[ch]) {
                node.children[ch] = new TrieNode();
            }
            node = node.children[ch];
        }
        node.isEnd = true;
    }
}

var replaceWords = function(dictionary, sentence) {
    const trie = new Trie();
    for (let word of dictionary) {
        trie.insert(word);
    }

    const words = sentence.split(' ');
    for (let i = 0; i < words.length; i++) {
        let node = trie.root;
        for (let j = 0; j < words[i].length; j++) {
            if (!node.children[words[i][j]]) {
                break;
            }
            node = node.children[words[i][j]];
            if (node.isEnd) {
                words[i] = words[i].substring(0, j + 1);
                break;
            }
        }
    }

    return words.join(' ');
};